gk/fib.c

```diff
@@ -0,0 +1,1440 @@
+{
+	eth_cache->filled = true;
+	eth_cache->stale = true;
+	rte_memcpy(&eth_cache->ip_addr, addr, sizeof(eth_cache->ip_addr));
+	eth_cache->eth_hdr.ether_type = addr->proto;
+	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
+	eth_cache->ref_cnt = 1;
+}
+
+/* Create a FIB entry for the gateway. */
+static struct gk_fib *
+create_fib_entry_for_gateway_locked(struct ipaddr *ip_addr, int prefix_len,
+	enum gk_fib_action action, struct ipaddr *gw_addr,
+	struct gatekeeper_if *iface, struct gk_config *gk_conf)
+{
+	int ret, fib_id;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	struct gk_fib *gw_fib;
+	struct ether_cache *eth_cache;
+	struct neighbor_hash_table *neigh_ht;
+
+	/* Find the neighbor FIB entry for this gateway. */
+	struct gk_fib *neigh_fib = find_fib_entry_for_neighbor_locked(
+		gw_addr, action, iface, gk_conf);
+	if (neigh_fib == NULL)
+		return NULL;
+
+	/* Find the Ethernet cached header entry for this gateway. */
+	neigh_ht = &neigh_fib->u.neigh;
+	eth_cache = neigh_get_ether_cache(neigh_ht, &gw_addr->ip);
+	if (eth_cache == NULL)
+		return NULL;
+
+	/* Find an empty FIB entry for the Gateway. */
+	fib_id = get_empty_fib_id(ip_addr->proto, gk_conf);
+	if (fib_id < 0)
+		return NULL;
+
+	if (ip_addr->proto == ETHER_TYPE_IPv4)
+		gw_fib = &ltbl->fib_tbl[fib_id];
+	else
+		gw_fib = &ltbl->fib_tbl6[fib_id];
+
+	/* Fills up the Gateway FIB entry for the IP prefix. */
+	gw_fib->action = action;
+	gw_fib->u.gateway.eth_cache = eth_cache;
+
+	if (!eth_cache->filled)
+		fill_up_ether_cache_locked(eth_cache, gw_addr, iface);
+	else {
+		/* TODO Add concurrency control. */
+		eth_cache->ref_cnt++;
+	}
+
+	ret = lpm_add_route(ip_addr, prefix_len, fib_id, ltbl);
+	if (ret < 0) {
+		eth_cache->ref_cnt--;
+		if (eth_cache->ref_cnt == 0)
+			neigh_del_ether_cache(neigh_ht, &gw_addr->ip);
+		initialize_fib_entry(gw_fib);
+		return NULL;
+	}
+
+	return gw_fib;
+}
+
+/*
+ * Initialize a gateway FIB entry.
+ * @gateway the gateway address informaiton.
+ * @ip_prefix the IP prefix,
+ * for which the gateway is responsible.
+ */
+static struct gk_fib *
+init_gateway_fib_locked(struct ip_prefix *ip_prefix, enum gk_fib_action action,
+	struct ipaddr *gw_addr , struct gk_config *gk_conf)
+{
+	struct gatekeeper_if *iface;
+
+	if (gw_addr->proto != ip_prefix->addr.proto) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since the gateway and its responsible IP prefix have different IP versions.\n");
+		return NULL;
+	}
+
+	if (action == GK_FWD_GATEWAY_FRONT_NET)
+		iface = &gk_conf->net->front;
+	else if (likely(action == GK_FWD_GATEWAY_BACK_NET))
+		iface = &gk_conf->net->back;
+	else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since it has invalid action %d.\n",
+			action);
+		return NULL;
+	}
+
+	/*
+	 * Create a Gateway FIB entry for the IP prefix,
+	 * for which the gateway is responsible.
+	 */
+	return create_fib_entry_for_gateway_locked(&ip_prefix->addr,
```

Comment:
`create_fib_entry_for_gateway_locked()` is only used here, its meaning is also dependent on this use, and bringing its code here won't create a huge function.